add: pomerium identity-aware access proxy

desimone opened the pull request:
#### What this PR does / why we need it:
This PR adds support for [Pomerium](https://github.com/pomerium/pomerium) an identity-aware access proxy. 

(edit): example install on GKE using this [script](https://github.com/pomerium/pomerium/blob/0fb9443e64dac5632533d80ff2afed24c3b88c86/scripts/helm_gke.sh)
https://asciinema.org/a/YcYC4iZLZi5kCCU5lQIWzFnhV 

#### Checklist
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Variables are documented in the README.md

desimone committed:
add: pomerium identity-aware access proxy

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -0,0 +1,111 @@
+# settings that are shared by all services
+config:
+  # secrets can be generated with `head -c32 /dev/urandom | base64`
+  sharedSecret: "LL3g/NoKkao7mTec2/wYSOZAki1VbR89q+kqbdFj0c4="
```
@desimone it would be good if it was possible to specify a secret name as an alternative to the one generated by the template. For example I am a user of sealed-secrets and thus I generate the secrets myself.

desimone replied:
Will do. Could you point me to an example? I know some Charts use `randAlphaNum` (and the like) but unfortunately that's not quite the same as trying to get an exactly 32 byte random key.

victornoel replied:
Sorry, maybe I wasn't clear. I meant to add a property `existingSecret` for example and if it is set to, say, `my-secret`, then secrets.yaml would generate nothing and the deployments would refer to `my-secret` instead of the one normally generated by `secrets.yaml`.
A bit like this: https://github.com/helm/charts/blob/master/stable/oauth2-proxy/values.yaml#L11

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -0,0 +1,111 @@
+  # default is to use a self-signed cert, should _NOT_ be used inproduction
+  key: "LS0tLS1CRUdJTiBFQyBQUklWQVRFIEtFWS0tLS0tCk1IY0NBUUVFSUtzZ3pUNDdBQXlzTkJNa1BJYnVJV1BodFEwN2RzWGx4WjB5Z3pjV3VXaVVvQW9HQ0NxR1NNNDkKQXdFSG9VUURRZ0FFLzMyY2c0YWpuWHpyT2pORjlXczROK2NyUTdwTGVWOGt6Y3R6cGN1S2F1KzAzVmZXRlRaTQpaTmh2cFQrUzZlS3QzSW1ycExsRTlKRFN6dDZwcUI0UVZRPT0KLS0tLS1FTkQgRUMgUFJJVkFURSBLRVktLS0tLQo="
+  cert: "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUJFekNCdWdJSkFLK3VPcmsyVzFRTk1Bb0dDQ3FHU000OUJBTUNNQkV4RHpBTkJnTlZCQU1NQm5WdWRYTmwKWkRBZ0Z3MHhPVEF6TWpBeU1UTTNNemRhR0E4ME56VTNNREl4TkRJeE16Y3pOMW93RVRFUE1BMEdBMVVFQXd3RwpkVzUxYzJWa01Ga3dFd1lIS29aSXpqMENBUVlJS29aSXpqMERBUWNEUWdBRS8zMmNnNGFqblh6ck9qTkY5V3M0Ck4rY3JRN3BMZVY4a3pjdHpwY3VLYXUrMDNWZldGVFpNWk5odnBUK1M2ZUt0M0ltcnBMbEU5SkRTenQ2cHFCNFEKVlRBS0 ...
+  policyFile: |-
```
@desimone same here, if I want to have my own `ConfigMap`, it could be neat to be able to specify a `ConfigMap` name instead of its content.

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -0,0 +1,111 @@
+service:
+  # Service type can be set to ClusterIP, NodePort or LoadBalancer.
+  type: ClusterIP
+  externalPort: 443
```
@desimone is it mandatory to use https or is it possible to expose a non-https proxy so that ssl termination can be handled by nginx-ingress for example? If yes, the chart should allow for it I believe

desimone replied:
TLS is a requirement for all communication. Both between services and behind the ingress. Termination can still occur but the connection between pomerium and the load balancer is still assumed to be encrypted (ideally, mutually authenticated). 

I will better document the process in https://github.com/pomerium/pomerium/pull/65

victornoel replied:
ah ok, that's what I was missing! And is this going to stay a hard requirement? We could imagine doing http when staying behind the ingress hypothetically, especially to avoid doing double work with the ingress doing also http.

desimone replied:
Totally fair question, and probably is something I should be doing a better job of explaining in the docs. 

So to give a little bit more more context here...  One of the reasons I started working on pomerium was that I felt that the assumptions baked into a perimeter security posture -- that "things behind the firewall were secure" --  was flawed. Reading the beyondcorp papers, and learning more about Google's aurora, and NSA breach reinforced the importance of adopting a zero-trust stance, even within your own (or cloud) datacenter, when it comes to service communication. Because of that, I feel the ideal is to mutually authenticate and encrypt between all services even behind the ingress. 

Of course, that decision has trade-offs. As you mentioned, we are doing double duty on our TLS-terminating load balancer. Also, we have to setup maintain PKI to handle the encryption. 

So on one hand, I totally understand wanting to do TLS termination, and gain the performance benefit and management ease of not having to handle TLS connections, but on the other, if pomerium is going to be a "zero-trust" beyondcorp-inspired solution, and not just an authenticating reverse proxy, it feels like mTLS is a cornerstone that. 

tl;dr; - I think the security win is worth the performance hit, and we can improve the UX of the mTLS so it isn't so much of a burden. But I am very open to feedback.

victornoel replied:
Thank for taking the time to explain, this makes total sense. And anyway, this is not blocking I believe :)

desimone replied:
I added some logic to bootstrap the certificate/ca process.

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -0,0 +1,111 @@
+      - admins
+
+authenticate:
+  name: authenticate
```
@desimone this doesn't seem to be used anywhere…

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -0,0 +1,111 @@
+    clientSecret: ""
+
+authorize:
+  name: authorize
```
@desimone this doesn't seem to be used anywhere…

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -0,0 +1,111 @@
+  name: authorize
+
+proxy:
+  name: proxy
```
@desimone this doesn't seem to be used anywhere…

victornoel reviewed (commented):
Hey @desimone here is a first batch of comment on the chart :)

victornoel commented on stable/pomerium/templates/proxy-deployment.yaml:
```diff
@@ -0,0 +1,146 @@
+        - name: AUTHENTICATE_SERVICE_URL
+          value: {{ .Values.proxy.authenticateServiceUrl }}
+        - name: AUTHORIZE_SERVICE_URL
+          value: {{ .Values.proxy.authorizeServiceUrl }}
```
@desimone it could be interesting for the chart to directly rely on  `AUTHENTICATE_INTERNAL_URL` and `AUTHORIZE_INTERNAL_URL` since they are deployed in the same cluster, but maybe it is on purpose? I remember having problem with that with buzzfeed-sso behind a nginx-ingress…

desimone replied:
I've added some code to infer the internal url if none is provided. Behind the ingress communication is especially tricky, and is basically requires the use of `AUTHENTICATE_INTERNAL_URL` and `AUTHORIZE_INTERNAL_URL` in kubernetes and docker-compose environments.

victornoel commented on stable/pomerium/templates/ingress.yaml:
```diff
@@ -0,0 +1,36 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ include "pomerium.fullname" }}
```
@desimone you are missing the `.` after the template name

desimone replied:
Oops wrong place.

victornoel commented on stable/pomerium/templates/ingress.yaml:
```diff
@@ -0,0 +1,36 @@
+        paths:
+          - path: {{ .Values.ingress.path }}
+            backend:
+              serviceName: {{ include "pomerium.fullname" }}
```
@desimone you are missing the `.` after the template name

victornoel replied:
also shouldn't this point to `{{ include "pomerium.fullname" . }}-proxy` instead or something like that?

desimone replied:
Good catch. I should have updated all instances to use `{{ template "pomerium.fullname" . }}-{{ .Values.SERVICE.name }}`

victornoel commented on stable/pomerium/templates/authenticate-deployment.yaml:
```diff
@@ -0,0 +1,155 @@
+          name: {{ template "pomerium.fullname" . }}-policy
+     
+      volumes:
+      - name: {{ template "pomerium.fullname" . }}-policy
```
@desimone this doesn't need to be prefixed by the fullname because it is local to the deployment

desimone replied:
Is the updated correct?

victornoel replied:
yep, exactly, and there is the same thing in authorize-deployment by the way

desimone committed:
Add logic to stand up dummy CA and self signed certs.
Add victornoel to owners.
Several other small fixes.

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
fix naming issue

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
s/ONWERS/OWNERS

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone commented on stable/pomerium/OWNERS:
```diff
@@ -0,0 +1,6 @@
+approvers:
+- desimone
+- victornoel
```
@victornoel Thank you for the help reviewing this (and other) PRs/ issues. You have a much better handle on helm than I do. Let me know if adding you here is okay.

victornoel replied:
@desimone here, that's nice :) For now I'm ok being reviewer but not approver, I won't have the time to dedicate to it

victornoel commented on stable/pomerium/templates/authenticate-deployment.yaml:
```diff
@@ -0,0 +1,150 @@
+            port: https
+            scheme: HTTPS
+        
+        
```
@desimone by the way, there a lot of empty line with useless whitespaces, maybe it would be cleaner to remove all of those

victornoel commented on stable/pomerium/templates/authenticate-deployment.yaml:
```diff
@@ -0,0 +1,150 @@
+{{ toYaml .Values.podAnnotations | indent 8 }}
+{{- end }}
+      labels:
+        app: {{ template "pomerium.name" . }}
```
@desimone there is a problem here with the labels of the deployments: there must be some label to explicitly differentiate this pod from the pods of the 2 other deployments.
Here in the pod `labels` and in the `matchLabels` above.
And this must be coherent with the service labels too!

desimone replied:
Great catch!

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -0,0 +1,144 @@
+      - admins
+tls:
+  generateTLS: true
+  certificateAuthority: ""
```
@desimone this doesn't seem to be used anywhere…

desimone replied:
I'm a little stuck on what to do with ingress right now, which is what I was going to use this for if not using external certs.

victornoel replied:
hehe, yes, usually the ingress is quite specific to each deployment. Personally I wrote a chart that depends upon the pomerium one and I generate the ingress as well as the policy file from a values file containing all the needed data to generate it :)

victornoel commented on stable/pomerium/templates/secret.yaml:
```diff
@@ -0,0 +1,40 @@
+{{- $ca := genCA "default-ca" 365 }}
+{{- $cn := default "example.com" .Values.tls.defaultCN }}
+  ca-cert: {{ $ca.Cert | b64enc | b64enc }}
+  ca-key: {{ $ca.Key | b64enc | b64enc }}
```
@desimone this is not used anywhere

desimone replied:
I'm not sure what best practice would be for setting up a dummy ca. Would you expect to be able to recover the key or would be fine with throwing it away? Of course the ideal case is to use "real" pki infrastructure but for one-off here I thought it may be better to keep it ? (to revoke?)

victornoel replied:
I didn't even think of that ^^ I just noticed that it wasn't used and thought that since it was not for prod, it wasn't very important. So that's a good question actually, but I have no answer :)

desimone replied:
I think it is fine to keep for a user generating a dummy ca, and gives them the flexibility to revoke, or extend signed certs later. But I should sufficiently warn users in the docs.

desimone committed:
updates

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
Mistakenly added policy to authenticate not authorize.

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

victornoel commented on stable/pomerium/templates/proxy-deployment.yaml:
```diff
@@ -53,41 +53,42 @@ spec:
         - name: POLICY_FILE
           value: /etc/pomerium/policy.yaml
-{{- end }}
-{{- if .Values.config.policy}}
```
@desimone you removed the conditional here and above, is that on purpose? I don't think the user will use both `POLICY` and `POLICY_FILE`, will he?

desimone replied:
I did because pomerium will always prefer the file, but that's undocumented behavior and is better to be explicit.

victornoel commented on stable/pomerium/templates/proxy-deployment.yaml:
```diff
@@ -107,17 +108,13 @@ spec:
       volumes:
       - name: policy
         configMap:
           name: {{ template "pomerium.fullname" . }}
```
@desimone you are missing something like `{{ if not .Values.config.existingConfig }}{{ include "pomerium.fullname" . }}{{ else }}{{ .Values.config.existingConfig }}{{ end }}` here I believe

victornoel commented on stable/pomerium/templates/proxy-deployment.yaml:
```diff
@@ -53,41 +53,42 @@ spec:
           valueFrom:
             secretKeyRef:
-              name: {{ template "pomerium.fullname" . }}
-              key: proxy-cert
```
@desimone you mistakenly removed the key here

victornoel commented on stable/pomerium/templates/proxy-deployment.yaml:
```diff
@@ -53,41 +53,42 @@ spec:
           valueFrom:
             secretKeyRef:
-              name: {{ template "pomerium.fullname" . }}
-              key: proxy-key
```
@desimone you mistakenly removed the key here

desimone replied:
Thank you!

victornoel commented on stable/pomerium/templates/proxy-deployment.yaml:
```diff
@@ -53,41 +53,42 @@ spec:
           valueFrom:
             secretKeyRef:
-              name: {{ template "pomerium.fullname" . }}
-              key: ca-cert
```
@desimone you mistakenly removed the key here

victornoel commented on stable/pomerium/templates/proxy-service.yaml:
```diff
@@ -1,7 +1,7 @@
-  name: {{ template "pomerium.fullname" . }}-proxy
+  name: {{ template "pomerium.fullname" . }}-{{ .Values.proxy.name }}
   labels:
     app: {{ template "pomerium.name" . }}
```
@desimone you forgot the `-{{ .Values.proxy.name }}`here

desimone committed:
Updates

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
improved ingress, change proxyRootDomain to rootDomain

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
updates

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
Autoconfigure what we can.
Fix issues pointed out by victor

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

victornoel commented on stable/pomerium/templates/proxy-deployment.yaml:
```diff
@@ -53,26 +55,22 @@ spec:
-          value: {{ template "pomerium.fullname" . }}-{{ .Values.authenticate.name }}
-{{- end }}
-{{- if .Values.proxy.authorizeUseInternalUrl }}
+          value: {{ default (printf "%s-%s.%s.svc.cluster.local" (include "pomerium.fullname" .) .Values.authenticate.name .Release.Namespace ) .Values.proxy.authenticateUseInternalUrl}}
```
@desimone is this working? If authenticateUseInternalUrl is set (to false or true), it should use its value… no?

victornoel replied:
@desimone ah but maybe you weren't planning for `authenticateUseInternalUrl` to be a boolean. So if it's an empty string, I think `default` will use the default alternative, not the empty string.

Also, if it's a string, I think it should be named `authenticateInternalUrl` (without the `Use` word).

desimone replied:
Good catch.

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -4,13 +4,11 @@
 # settings that are shared by all services
 config:
   # routes under this wildcard comain are handled by pomerium
```
@desimone you wrote `comain` here :P

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -35,83 +33,64 @@ config:
+    # {}
+    # if using GKE's build int load balancer
+    cloud.google.com/app-protocols: '{"https":"HTTPS"}'
+    kubernetes.io/ingress.allow-http: "false"
```
@desimone should we impose GKE to the chart user by default?

desimone replied:
Yeah this was left in from testing thanks. 

Though it's required for GKE, from looking at other charts, it looks like ingress is left as an exercise to the user because they vary so much. I think it would be helpful to keep it commented though.

victornoel replied:
yep, and also we could add some example for nginx-ingress commented out:
```
kubernetes.io/ingress.class: nginx
nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
# authenticate sends big headers or something
nginx.ingress.kubernetes.io/proxy-buffer-size: "16k"
```

desimone replied:
Good call. The buffer size thing is a common issue. 

https://github.com/pomerium/nginx-proxy/commit/0e12ab7ce4f15939db25502d882c6be1990abff9

victornoel commented on stable/pomerium/values.yaml:
```diff
@@ -0,0 +1,130 @@
+# settings that are shared by all services
+config:
+  # routes under this wildcard comain are handled by pomerium
+  domain: corp.pomerium.io
```
@desimone you renamed it here but not in many other places where it is still named `rootDomain`.

desimone committed:
updates

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
tidy up values

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
updates

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
pomerium : remove unnecessary spaces

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
tag release to v0.0.3

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

mattfarina commented on stable/pomerium/templates/authenticate-deployment.yaml:
```diff
@@ -0,0 +1,141 @@
+    app: {{ template "pomerium.name" . }}-{{ .Values.authenticate.name }}
+    chart: {{ template "pomerium.chart" . }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
```
The common labels changed awhile back due to recommendations from Kubernetes. Can you update them per the docs at https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md#metadata. This applies to all the objects

mattfarina commented on stable/pomerium/templates/authenticate-deployment.yaml:
```diff
@@ -0,0 +1,141 @@
+    chart: {{ template "pomerium.chart" . }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ template "pomerium.fullname" . }}-{{ .Values.authenticate.name }}
```
The name field in Kubernetes is limited to 63 characters. This is why the `fullname` function (see the helpers) truncates the name that way. By adding something to the end there is the possibility the name will be to long and cause install errors. We have encountered these which is why things are limited now. Can you please use functions for the various names and limit their length to avoid Kubernetes errors.

desimone replied:
Makes complete sense. Just a FYI, the docs do give what I had before as an example. Might be worth mentioning the character limit there as well.

>> The resource name should get the component as suffix (e. g. name: {{ include "myapp.fullname" . }}-server).

mattfarina commented on stable/pomerium/values.yaml:
```diff
@@ -0,0 +1,133 @@
+image:
+  repository: "pomerium/pomerium"
+  tag: "v0.0.3"
+  pullPolicy: "Always"
```
When pinned to a tag the `pullPolicy` can be `ifNotPresent` so that caching can be used locally. Can you please update.

desimone replied:
Yes, will do.

desimone committed:
Refactor labels to use updated Kubernetes keys.
Use helper templates for names to avoid going over 63 chars.
Add component label since this is a multi-service app.
Replace name with "nameOverride and fullnameOverride"

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
fix automatic certificate generation

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
fix service name fall through

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>

desimone committed:
remove unused variable

Signed-off-by: Bobby DeSimone <bobbydesimone@gmail.com>
